### ISSUE TITLE
NullReferenceException in JsonClientCache

### ISSUE THREAD

**Issue by tompetko** [sonvister/Binance#75]
Setting `Client` property of `ISymbolStatisticsCache` leads to unwanted subscription and causes  `NullReferenceException` in `ClientCallback` method of `JsonClientCache`, because `_buffer` is null because `Subscribe` was not called before setting Client property of cache object.

I believe the problem is in `SubscribeToClient` method of `JsonClientCache`. When `Client` property is set, LinkTo method calls SubscribeToClient that causes subscription to all markets. This is why Subscribe on `SymbolStatisticsCache` was never called, thus `_symbols` field is in it's default state - empty list and SubscribeToClient thinks it should subscribe to all markets. 

```
var serviceProvider = new ServiceCollection()
    .AddBinance() // add default Binance services.
    .AddLogging(builder => builder.SetMinimumLevel(LogLevel.Trace))
    .BuildServiceProvider();

var symbolStatsClientManager = serviceProvider.GetService<ISymbolStatisticsWebSocketClientManager>();

symbolStatsClientManager.Controller.Error += Controller_Error;
symbolStatsClientManager.StatisticsUpdate += SymbolStatsClient_StatisticsUpdate;

var symbolStatsCache = serviceProvider.GetService<ISymbolStatisticsCache>();
symbolStatsCache.Client = this.symbolStatsClientManager;
```

Anyway, as I'm looking at the `JsonClientCache`, it would be nice to also handle `NullReferenceException` in `OnUnsubscribe` method after calling `Unsubsribe` when no subscription was made or after calling it multiple times.

Version: *0.2.0-alpha39*

**Comment by sonvister** [comment]
@tompetko, thanks for reporting the issues. I am able to reproduce the following in 0.2.0-alpha39:

**Issues:**

1. `SymbolStatisticsCache` subscribes linked `ISymbolStatisticsClient` to all symbols when cache is unsubscribed.
    ```C#
    var client = services.GetService<ISymbolStatisticsWebSocketClient>();
    var cache = services.GetService<ISymbolStatisticsCache>();

    cache.Client = client;
    ```

2. `NullReferenceException` when calling calling `SymbolStatisticsCache.Unsubscribe` before `SymbolStatisticsCache.Subscribe` or after `SymbolStatisticsCache.Unsubscribe`.
    ```C#
    var cache = services.GetService<ISymbolStatisticsCache>();

    cache.Unsubscribe();
    ```

**Workarounds:**
1. Use `ISymbolStatisticsWebSocketCacheManager`.
2. Only call `Unsubscribe` once *after* calling `Subscribe`.

**Comment by tompetko** [comment]
Thanks for fast response and fix @sonvister. Are you planning to deploy it any time soon?

**Comment by sonvister** [comment]
@tompetko, no problem. The changes are included in [0.2.0-alpha40](https://github.com/sonvister/Binance/releases/tag/v0.2.0-alpha40). There are *some* other changes too, so I'll let you close this issue (*to confirm all is resolved*). Thanks.

**Comment by tompetko** [comment]
Works great @sonvister. Thank you for extra-fast fix. Closing issue.
